File: sdk/typescript/src/dpack.ts
/**
 * D:PACK/BLK resolve — read-only access to DBLK v1 binaries.
 *
 * Resolves a single key from a block-compressed corpus by decompressing
 * only the containing block.  Uses fzstd (82KB, pure JS, zero native deps)
 * for zstd decompression.
 *
 * DBLK v1 format: see OSMP-SPEC-v1.md §10.4.1
 *
 * Dependency: fzstd (npm install fzstd)
 *   Dict-free DBLK binaries only.  fzstd does not support external
 *   trained dictionaries.  Binaries built with BlockCompressor(use_dict=False)
 *   in the Python SDK are fully compatible.  Dict-free binaries are 0.1-1.5%
 *   larger than dict variants; the tradeoff is zero native dependencies in
 *   the TypeScript SDK.
 */

import { decompress } from "fzstd";

// ── constants ───────────────────────────────────────────────────────

const DBLK_MAGIC = 0x44424c4b; // "DBLK"
const DBLK_HEADER_SIZE = 24;
const DBLK_FIRST_CODE_SIZE = 32;
const DBLK_BTABLE_ENTRY_SIZE = 44; // 32 + 4 + 4 + 2 + 2
const MAX_DECOMPRESS_SIZE = 40960; // 32KB target + 8KB headroom

// ── types ───────────────────────────────────────────────────────────

export interface BlkStats {
  totalBytes: number;
  headerBytes: number;
  btableBytes: number;
  dictBytes: number;
  blockDataBytes: number;
  blockCount: number;
}

// ── header parsing ──────────────────────────────────────────────────

interface DblkHeader {
  version: number;
  flags: number;
  blockCount: number;
  dictOffset: number;
  dictSize: number;
  blocksOffset: number;
}

function parseHeader(data: Uint8Array): DblkHeader {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const magic = view.getUint32(0);
  if (magic !== DBLK_MAGIC) {
    throw new Error(`Not a DBLK binary (magic: 0x${magic.toString(16)})`);
  }
  return {
    version: view.getUint16(4),
    flags: view.getUint16(6),
    blockCount: view.getUint32(8),
    dictOffset: view.getUint32(12),
    dictSize: view.getUint32(16),
    blocksOffset: view.getUint32(20),
  };
}

// ── block table binary search ───────────────────────────────────────

function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

function findBlock(data: Uint8Array, hdr: DblkHeader, code: string): number {
  const codeBytes = new TextEncoder().encode(code);
  let lo = 0;
  let hi = hdr.blockCount - 1;
  let result = 0;

  while (lo <= hi) {
    const mid = (lo + hi) >>> 1;
    const off = DBLK_HEADER_SIZE + mid * DBLK_BTABLE_ENTRY_SIZE;

    // extract first_code, strip null padding
    let fcLen = DBLK_FIRST_CODE_SIZE;
    while (fcLen > 0 && data[off + fcLen - 1] === 0) fcLen--;
    const fc = data.subarray(off, off + fcLen);

    if (compareBytes(fc, codeBytes) <= 0) {
      result = mid;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  return result;
}

// ── block decompression + search ────────────────────────────────────

function decompressBlock(
  data: Uint8Array,
  hdr: DblkHeader,
  blockIdx: number
): Uint8Array {
  const entryOff = DBLK_HEADER_SIZE + blockIdx * DBLK_BTABLE_ENTRY_SIZE;
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const blkOffset = view.getUint32(entryOff + DBLK_FIRST_CODE_SIZE);
  const blkCsize = view.getUint32(entryOff + DBLK_FIRST_CODE_SIZE + 4);

  const start = hdr.blocksOffset + blkOffset;
  const compressed = data.subarray(start, start + blkCsize);

  return decompress(compressed);
}

function searchBlock(decompressed: Uint8Array, code: string): string | null {
  const text = new TextDecoder().decode(decompressed);
  const lines = text.split("\n");
  for (const line of lines) {
    const tab = line.indexOf("\t");
    if (tab > 0 && line.slice(0, tab) === code) {
      return line.slice(tab + 1);
    }
  }
  return null;
}

// ── public API ──────────────────────────────────────────────────────

/**
 * Resolve a single key from a DBLK binary.
 *
 * Decompresses only the block containing the target key.
 * When the 32-byte first_code truncation causes the binary search
 * to overshoot, the previous block is checked as a fallback.
 *
 * @param data  Complete DBLK binary as Uint8Array
 * @param code  Key to look up (MDR token or type name)
 * @returns     SAL description text, or null if not found
 */
export function resolveBlk(data: Uint8Array, code: string): string | null {
  const hdr = parseHeader(data);

  if (hdr.flags & 1 && hdr.dictSize > 0) {
    throw new Error(
      "This DBLK binary uses a trained dictionary. " +
        "The TypeScript SDK requires dict-free binaries " +
        "(built with BlockCompressor(use_dict=False) in Python)."
    );
  }

  const blkIdx = findBlock(data, hdr, code);

  const raw = decompressBlock(data, hdr, blkIdx);
  const result = searchBlock(raw, code);
  if (result !== null) return result;

  // Truncation fallback: try previous block
  if (blkIdx > 0) {
    const prev = decompressBlock(data, hdr, blkIdx - 1);
    return searchBlock(prev, code);
  }

  return null;
}

/**
 * Return structural statistics for a DBLK binary.
 */
export function statsBlk(data: Uint8Array): BlkStats {
  const hdr = parseHeader(data);
  const btableBytes = hdr.blockCount * DBLK_BTABLE_ENTRY_SIZE;
  return {
    totalBytes: data.length,
    headerBytes: DBLK_HEADER_SIZE,
    btableBytes,
    dictBytes: hdr.dictSize,
    blockDataBytes: data.length - hdr.blocksOffset,
    blockCount: hdr.blockCount,
  };
}

File: sdk/go/osmp/dpack.go
// Package osmp provides the OSMP protocol implementation.
//
// This file implements D:PACK/BLK resolve: read-only access to DBLK v1
// binaries.  Resolves a single key by decompressing only the containing
// block.
//
// Dependency: github.com/klauspost/compress/zstd (decoder only).
// The compiled binary includes only the decoder path.  The full module
// source is large but Go compiles only imported packages.
//
// Supports both dict-free and dict-compressed DBLK binaries.
// Dict-free binaries (flags bit 0 = 0) require no dictionary setup.
// Dict binaries (flags bit 0 = 1) load the trained dictionary from
// the binary header.
//
// DBLK v1 format: see OSMP-SPEC-v1.0.2.md section 10.4.1
package osmp

import (
	"bytes"
	"encoding/binary"
	"errors"
	"strings"

	"github.com/klauspost/compress/zstd"
)

// DBLK format constants
const (
	dblkMagic          = 0x44424c4b // "DBLK"
	dblkHeaderSize     = 24
	dblkFirstCodeSize  = 32
	dblkBtableEntrySize = 44
	maxDecompressSize  = 40960 // 32KB target + 8KB headroom
)

// DblkStats holds structural statistics for a DBLK binary.
type DblkStats struct {
	TotalBytes     int
	HeaderBytes    int
	BtableBytes    int
	DictBytes      int
	BlockDataBytes int
	BlockCount     int
}

type dblkHeader struct {
	version      uint16
	flags        uint16
	blockCount   uint32
	dictOffset   uint32
	dictSize     uint32
	blocksOffset uint32
}

func parseDblkHeader(data []byte) (dblkHeader, error) {
	if len(data) < dblkHeaderSize {
		return dblkHeader{}, errors.New("data too short for DBLK header")
	}
	magic := binary.BigEndian.Uint32(data[0:4])
	if magic != dblkMagic {
		return dblkHeader{}, errors.New("not a DBLK binary (bad magic)")
	}
	return dblkHeader{
		version:      binary.BigEndian.Uint16(data[4:6]),
		flags:        binary.BigEndian.Uint16(data[6:8]),
		blockCount:   binary.BigEndian.Uint32(data[8:12]),
		dictOffset:   binary.BigEndian.Uint32(data[12:16]),
		dictSize:     binary.BigEndian.Uint32(data[16:20]),
		blocksOffset: binary.BigEndian.Uint32(data[20:24]),
	}, nil
}

func findDblkBlock(data []byte, hdr dblkHeader, code string) int {
	codeBytes := []byte(code)
	lo, hi, result := 0, int(hdr.blockCount)-1, 0
	for lo <= hi {
		mid := (lo + hi) / 2
		off := dblkHeaderSize + mid*dblkBtableEntrySize

		// extract first_code, strip null padding
		fc := data[off : off+dblkFirstCodeSize]
		fcLen := dblkFirstCodeSize
		for fcLen > 0 && fc[fcLen-1] == 0 {
			fcLen--
		}
		fc = fc[:fcLen]

		if bytes.Compare(fc, codeBytes) <= 0 {
			result = mid
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}
	return result
}

func decompressDblkBlock(data []byte, hdr dblkHeader, blockIdx int, decoder *zstd.Decoder) ([]byte, error) {
	entryOff := dblkHeaderSize + blockIdx*dblkBtableEntrySize
	blkOffset := binary.BigEndian.Uint32(data[entryOff+dblkFirstCodeSize : entryOff+dblkFirstCodeSize+4])
	blkCsize := binary.BigEndian.Uint32(data[entryOff+dblkFirstCodeSize+4 : entryOff+dblkFirstCodeSize+8])

	start := int(hdr.blocksOffset) + int(blkOffset)
	compressed := data[start : start+int(blkCsize)]

	buf := make([]byte, 0, maxDecompressSize)
	return decoder.DecodeAll(compressed, buf)
}

func searchDblkBlock(decompressed []byte, code string) (string, bool) {
	text := string(decompressed)
	for _, line := range strings.Split(text, "\n") {
		idx := strings.Index(line, "\t")
		if idx > 0 && line[:idx] == code {
			return line[idx+1:], true
		}
	}
	return "", false
}

// ResolveBlk resolves a single key from a DBLK binary.
//
// Decompresses only the block containing the target key.
// When the 32-byte first_code truncation causes the binary search
// to overshoot, the previous block is checked as a fallback.
//
// Returns the SAL description text and nil error, or empty string
// and nil if the key is not found.  Returns an error only for
// structural problems (bad magic, decompression failure).
func ResolveBlk(data []byte, code string) (string, error) {
	hdr, err := parseDblkHeader(data)
	if err != nil {
		return "", err
	}

	// Build decoder with or without dictionary
	var opts []zstd.DOption
	if hdr.flags&1 != 0 && hdr.dictSize > 0 {
		dictBytes := data[hdr.dictOffset : hdr.dictOffset+hdr.dictSize]
		opts = append(opts, zstd.WithDecoderDicts(dictBytes))
	}
	decoder, err := zstd.NewReader(nil, opts...)
	if err != nil {
		return "", err
	}
	defer decoder.Close()

	blkIdx := findDblkBlock(data, hdr, code)

	raw, err := decompressDblkBlock(data, hdr, blkIdx, decoder)
	if err != nil {
		return "", err
	}
	if result, ok := searchDblkBlock(raw, code); ok {
		return result, nil
	}

	// Truncation fallback: try previous block
	if blkIdx > 0 {
		raw, err = decompressDblkBlock(data, hdr, blkIdx-1, decoder)
		if err != nil {
			return "", err
		}
		if result, ok := searchDblkBlock(raw, code); ok {
			return result, nil
		}
	}

	return "", nil
}

// BlkStats returns structural statistics for a DBLK binary.
func BlkStats(data []byte) (*DblkStats, error) {
	hdr, err := parseDblkHeader(data)
	if err != nil {
		return nil, err
	}
	btable := int(hdr.blockCount) * dblkBtableEntrySize
	return &DblkStats{
		TotalBytes:     len(data),
		HeaderBytes:    dblkHeaderSize,
		BtableBytes:    btable,
		DictBytes:      int(hdr.dictSize),
		BlockDataBytes: len(data) - int(hdr.blocksOffset),
		BlockCount:     int(hdr.blockCount),
	}, nil
}
